File: sections/50_advanced_algorithms_mixed_contest/problems/l_profitable_disjoint_routes/solution.py
from collections import deque
import sys
def main():
    d=list(map(int,sys.stdin.buffer.read().split()))
    if not d:return
    n,m,k=d[:3];g=[[]for _ in range(n)];j=3
    def add(u,v,c):
        g[u].append([v,len(g[v]),1,c]);g[v].append([u,len(g[u])-1,0,-c])
    for _ in range(m):u,v,p=d[j:j+3];j+=3;add(u-1,v-1,-p)
    flow=0;cost=0;inf=10**30
    while flow<k:
        dist=[inf]*n;pv=[-1]*n;pe=[-1]*n;inside=[False]*n;dist[0]=0;q=deque([0]);inside[0]=True
        while q:
            u=q.popleft();inside[u]=False
            for i,e in enumerate(g[u]):
                v,rev,cap,c=e
                if cap and dist[v]>dist[u]+c:
                    dist[v]=dist[u]+c;pv[v]=u;pe[v]=i
                    if not inside[v]:inside[v]=True;q.append(v)
        if dist[-1]==inf:break
        flow+=1;cost+=dist[-1];v=n-1
        while v:
            u=pv[v];i=pe[v];rev=g[u][i][1];g[u][i][2]-=1;g[v][rev][2]+=1;v=u
    print(-cost if flow==k else 'IMPOSSIBLE')
if __name__=="__main__":main()

File: sections/50_advanced_algorithms_mixed_contest/problems/l_profitable_disjoint_routes/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
struct E {
  int to, rev, cap;
  ll cost;
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, m, k;
  if (!(cin >> n >> m >> k))
    return 0;
  vector<vector<E>> g(n);
  auto add = [&](int u, int v, ll c) {
    g[u].push_back({v, (int)g[v].size(), 1, c});
    g[v].push_back({u, (int)g[u].size() - 1, 0, -c});
  };
  while (m--) {
    int u, v;
    ll p;
    cin >> u >> v >> p;
    add(u - 1, v - 1, -p);
  }
  ll cost = 0;
  int flow = 0;
  const ll INF = 4e18;
  while (flow < k) {
    vector<ll> d(n, INF);
    vector<int> pv(n), pe(n);
    vector<char> in(n);
    queue<int> q;
    d[0] = 0;
    q.push(0);
    in[0] = 1;
    while (!q.empty()) {
      int u = q.front();
      q.pop();
      in[u] = 0;
      for (int i = 0; i < (int)g[u].size(); i++) {
        E &e = g[u][i];
        if (e.cap && d[e.to] > d[u] + e.cost) {
          d[e.to] = d[u] + e.cost;
          pv[e.to] = u;
          pe[e.to] = i;
          if (!in[e.to])
            in[e.to] = 1, q.push(e.to);
        }
      }
    }
    if (d[n - 1] == INF)
      break;
    cost += d[n - 1];
    flow++;
    for (int v = n - 1; v; v = pv[v]) {
      E &e = g[pv[v]][pe[v]];
      e.cap--;
      g[v][e.rev].cap++;
    }
  }
  if (flow < k)
    cout << "IMPOSSIBLE\n";
  else
    cout << -cost << '\n';
}
